### PR TITLE
Add conformance tests for connect-node

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Add `connect-node-conformance` package with server implementation. Based on the [reference server implementation](https://github.com/connectrpc/conformance/tree/v1/internal/app/referenceserver).

Tested with the following conformance config:
```yaml
features:
  versions:
    - HTTP_VERSION_1
    - HTTP_VERSION_2
  protocols:
    - PROTOCOL_CONNECT
    - PROTOCOL_GRPC
    - PROTOCOL_GRPC_WEB
  codecs:
    - CODEC_PROTO
    - CODEC_JSON
  compressions:
    - COMPRESSION_IDENTITY
    - COMPRESSION_GZIP
    - COMPRESSION_BR
  supportsTls: true
  supportsTlsClientCerts: true
  supportsConnectGet: true
  supportsHalfDuplexBidiOverHttp1: true
  requiresConnectVersionHeader: false
  supportsMessageReceiveLimit: true

```

### CODE REVIEW

File: packages/connect-node-conformance/src/protocol.ts
Comment (timostamm) [inline packages/connect-node-conformance/src/protocol.ts:22]:
```suggestion
  // The ConnectError constructor accepts messages for details.
```

**Review by timostamm — COMMENTED** [review]
I did not take a close look at the server implementation, but this looks solid 👍

Left one comment about error detail debug values. 

The readme for this new package needs more details, and the test run need to be wired up in make. I guess it only makes sense to do that after conformance v1 has been released. 

The old crosstests need to be removed from this repo. I guess this is only feasible after https://github.com/connectrpc/connect-es/pull/890 (we'll also want to run a Node fetch client, a browser fetch client, and at least retain the browserstack tests).

**Comment by srikrsna-buf** [comment]
> Still looking solid to me 🙂 Would be great to have one of the conformance authors take a look, though.
> 
> Doesn't seem great to merge before conformance is released. What's your take?

Agree on conformance authors taking a look, about merging before v1 we are only merging to a branch so that I can continue to work on environments

**Comment by srikrsna-buf** [comment]
@timostamm @smaye81 Added the test to CI. This is now good to go

**Review by timostamm — APPROVED** [review]
Would be nice to remove the dep on `fflate` as noted in https://github.com/connectrpc/connect-es/pull/886#discussion_r1410640486.

LGTM from me, but let's have @smaye81 take a look.

File: packages/connect-node-conformance/src/conformance.ts
Comment (jhump) [reply packages/connect-node-conformance/src/conformance.ts:29]: This cargo-culted from other release configs, where "tgz" is the norm, and "zip" is provided for Windows since that's the only format it supports out-of-the-box (users would have to download a separate tool or install wsl for a "tgz" file).

Does this have to be downloaded via JS code? I guess I was thinking this is easiest with a simple shell script that uses curl or wget and tar. The suffix of the file is designed to work with the output of uname.

FWIW, this appears to also be how we do releases of the CLI: https://github.com/bufbuild/buf/releases/tag/v1.28.1

**Comment by smaye81** [comment]
Sorry for the delay @srikrsna-buf. Left some comments.

File: packages/connect-node-conformance/src/conformance.ts
Comment (smaye81) [inline packages/connect-node-conformance/src/conformance.ts:102]: nit: typo on `Artifact`

File: packages/connect-node-conformance/src/conformance.ts
Comment (smaye81) [reply packages/connect-node-conformance/src/conformance.ts:29]: I'll leave it up to you guys on how to handle it, but I agree with @jhump that a shell script pulling down the binary would be the easiest. I understand the idiomatic thing, but since this is more of an ancillary process and not involved in the actual logic of Connect-ES maybe a shell script is ok? At least maybe pulling down the tarball and un-tarring it?

Comment (timostamm) [reply packages/connect-node-conformance/src/conformance.ts:29]: Shell scripts in Connect-ES, Protobuf-ES, and other repositories targeting TS and JS are not okay. They are a barrier for most users.

Comment (srikrsna-buf) [reply packages/connect-node-conformance/src/conformance.ts:29]: > `fflate` is a great package, but can we use `node:zlib` here, since we already use it?

I don't think we can use zlib to handle zip archives. But I removed the zlib usage to avoid using two different packages for decompression.

Comment (timostamm) [reply packages/connect-node-conformance/src/conformance.ts]: Right, I [remember now](https://github.com/timostamm/protobuf-ts/blob/6cdc5cb27c2fef0faf6a0d79fe1ede27b7335a54/packages/protoc/util.js#L328) :joy:

File: packages/connect-node-conformance/conformance.yaml
Comment (smaye81) [inline packages/connect-node-conformance/conformance.yaml:1]: Minor nits that can help clean this up:

* You can omit `protocols` and `codecs` since all options are supported (if absent, all are assumed)
* You can also omit `supportsTls`, `supportsConnectGet`, `requiresConnectVersionHeader`, and `supportsMessageReceiveLimit` since you have them listed as the defaults and if absent, the defaults are assumed.

**Review by smaye81 — APPROVED** [review]

File: packages/connect-node-conformance/conformance.yaml
Comment (srikrsna-buf) [reply packages/connect-node-conformance/conformance.yaml:1]: I did think of omitting the defaults but it just felt easier to parse and see what is included just by looking at the config. I'll omit them in a separate PR.

**Pull request merged**
